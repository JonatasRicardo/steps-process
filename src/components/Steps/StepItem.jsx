import React from 'react';
import PropTypes from 'prop-types';
import { faCheck } from '@fortawesome/free-solid-svg-icons';

import stepIcon from './img/step-icon.png';
import './steps.scss';

const propTypes = {
    title: PropTypes.string,
    statusName: PropTypes.string,
    statusDate: PropTypes.string,
    currentStep: PropTypes.number,
    stepIndex: PropTypes.number,
}

const defaultProps = {
    currentStep: -1,
    stepIndex: 0,
}

function StepItem(props) {
    const {
        currentStep,
        stepIndex,
        title,
        statusName,
        statusDate,
    } = props;

    const done = currentStep >= stepIndex;
    const doneClass = currentStep >= stepIndex ? 'is-done' : '';
    const nextDoneClass = currentStep > stepIndex ? 'is-next-done' : '';

    return (
        <div role="listitem" className={`step-item ${doneClass} ${nextDoneClass}`} data-testid="step-item">
            <span className="step-item__icon">
                <span>
                    {done && <img src={stepIcon} alt="step done" />}
                    <span className="step-item__index">{stepIndex}</span>
                </span>
            </span>
            {
                title &&
                <div className="step-item__title">
                    {title}
                </div>
            }
            <div className="step-item__status">
                {
                    statusName &&
                    <div className="step-item__status-name">
                        {statusName}
                    </div>
                }
                {
                    statusDate &&
                    <div className="step-item__status-date">
                        {statusDate}
                    </div>
                }
            </div>
        </div>
    );
}

StepItem.propTypes = propTypes;
StepItem.defaultProps = defaultProps;

export default StepItem;