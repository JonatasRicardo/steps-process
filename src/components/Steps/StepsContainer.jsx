import React from 'react';
import PropTypes from 'prop-types';
import './steps.scss';

const propTypes = {
    currentStep: PropTypes.number,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]),
}

const defaultProps = {
    currentStep: -1,
    stepIndex: 0,
}

function StepsContainer(props) {
    const { children, currentStep } = props;

    return (
        <div className="steps-container" role="list" data-testid="steps-container">
            {React.Children.map(children, (child, index) => {
                return React.cloneElement(child, {
                    currentStep,
                    stepIndex: index+1,
                })
            })}
        </div>
    )
}

StepsContainer.propTypes = propTypes;
StepsContainer.defaultProps = defaultProps;

export default StepsContainer;