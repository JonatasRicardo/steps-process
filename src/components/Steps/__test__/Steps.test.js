import React from 'react';
import { render } from '@testing-library/react';
import StepItem from '../StepItem';
import StepsContainer from '../StepsContainer';

  it('renders Steps components', () => {
    const { getByTestId, getAllByTestId } = render(
      <StepsContainer>
        <StepItem title="Execute" />
        <StepItem title="Concluded" />
        <StepItem title="Closed" />
      </StepsContainer>
    );
    const stepsContainerElement = getByTestId('steps-container');
    const stepItemElements = getAllByTestId('step-item');
    expect(stepsContainerElement).toBeDefined();
    expect(stepItemElements).toHaveLength(3)
  });

  it('Current Step classes', () => {
    const { getByTestId } = render(
      <StepsContainer currentStep={2}>
        <StepItem title="Execute" />
        <StepItem title="Concluded" />
        <StepItem title="Closed" />
      </StepsContainer>
    );
    const stepsContainerElement = getByTestId('steps-container');
    expect(stepsContainerElement.children[0]).toHaveClass('is-next-done');
    expect(stepsContainerElement.children[1]).toHaveClass('is-done');
    expect(stepsContainerElement.children[1]).not.toHaveClass('is-next-done');
    expect(stepsContainerElement.children[2]).not.toHaveClass('is-done');
    expect(stepsContainerElement.children[2]).not.toHaveClass('is-next-done');
  });