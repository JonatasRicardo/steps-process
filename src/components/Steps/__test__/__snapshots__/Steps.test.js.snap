// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepsContainer><StepItem/></StepsContainer> Snapshot 1`] = `
<div
  class="steps-container"
  data-testid="steps-container"
  role="list"
>
  <div
    class="step-item is-done is-next-done"
    data-testid="step-item"
    role="listitem"
  >
    <span
      class="step-item__icon"
    >
      <span>
        <img
          alt="step done"
          src="step-icon.png"
        />
        <span
          class="step-item__index"
        >
          1
        </span>
      </span>
    </span>
    <div
      class="step-item__title"
    >
      Draft
    </div>
    <div
      class="step-item__status"
    >
      <div
        class="step-item__status-name"
      >
        Created on
      </div>
      <div
        class="step-item__status-date"
      >
        01/31/2020
      </div>
    </div>
  </div>
  <div
    class="step-item is-done is-next-done"
    data-testid="step-item"
    role="listitem"
  >
    <span
      class="step-item__icon"
    >
      <span>
        <img
          alt="step done"
          src="step-icon.png"
        />
        <span
          class="step-item__index"
        >
          2
        </span>
      </span>
    </span>
    <div
      class="step-item__title"
    >
      Pending Approval
    </div>
    <div
      class="step-item__status"
    >
      <div
        class="step-item__status-name"
      >
        Approval sent
      </div>
      <div
        class="step-item__status-date"
      >
        02/14/2020
      </div>
    </div>
  </div>
  <div
    class="step-item is-done is-next-done"
    data-testid="step-item"
    role="listitem"
  >
    <span
      class="step-item__icon"
    >
      <span>
        <img
          alt="step done"
          src="step-icon.png"
        />
        <span
          class="step-item__index"
        >
          3
        </span>
      </span>
    </span>
    <div
      class="step-item__title"
    >
      Approved
    </div>
    <div
      class="step-item__status"
    >
      <div
        class="step-item__status-name"
      >
        Abandoned on
      </div>
      <div
        class="step-item__status-date"
      >
        02/20/2020
      </div>
    </div>
  </div>
  <div
    class="step-item is-done is-next-done"
    data-testid="step-item"
    role="listitem"
  >
    <span
      class="step-item__icon"
    >
      <span>
        <img
          alt="step done"
          src="step-icon.png"
        />
        <span
          class="step-item__index"
        >
          4
        </span>
      </span>
    </span>
    <div
      class="step-item__title"
    >
      Execute
    </div>
    <div
      class="step-item__status"
    />
  </div>
  <div
    class="step-item is-done is-next-done"
    data-testid="step-item"
    role="listitem"
  >
    <span
      class="step-item__icon"
    >
      <span>
        <img
          alt="step done"
          src="step-icon.png"
        />
        <span
          class="step-item__index"
        >
          5
        </span>
      </span>
    </span>
    <div
      class="step-item__title"
    >
      Concluded
    </div>
    <div
      class="step-item__status"
    />
  </div>
  <div
    class="step-item is-done "
    data-testid="step-item"
    role="listitem"
  >
    <span
      class="step-item__icon"
    >
      <span>
        <img
          alt="step done"
          src="step-icon.png"
        />
        <span
          class="step-item__index"
        >
          6
        </span>
      </span>
    </span>
    <div
      class="step-item__title"
    >
      Closed
    </div>
    <div
      class="step-item__status"
    />
  </div>
</div>
`;
